File: mobile-android/app/src/main/java/ph/attendance/domain/Reporting.kt
package ph.attendance.domain

import kotlin.math.roundToInt

data class StatusTally(val present: Int = 0, val late: Int = 0, val absent: Int = 0) {
    val total: Int get() = present + late + absent

    /**
     * Share of sessions the student actually turned up for. Late counts as attending — it is a
     * punctuality problem, not an absence, and conflating the two was a real reporting error in
     * the original app.
     */
    val rate: Float get() = if (total == 0) 0f else (present + late).toFloat() / total
}

fun List<AttendanceRecord>.tally(): StatusTally {
    var present = 0
    var late = 0
    var absent = 0
    for (record in this) {
        when (record.status) {
            AttendanceStatus.PRESENT -> present++
            AttendanceStatus.LATE -> late++
            AttendanceStatus.ABSENT -> absent++
        }
    }
    return StatusTally(present, late, absent)
}

fun formatRate(rate: Float): String = "${(rate * 100).roundToInt()}%"

data class StudentSummary(
    val student: Student,
    val counts: StatusTally,
) {
    val displayName: String get() = student.displayName
    val sessions: Int get() = counts.total
    val rate: Float get() = counts.rate
}

/**
 * Attendance per student over an inclusive date range.
 *
 * Students with no records are still listed, with a zero tally. A student who never appears in the
 * data is exactly the one a coordinator needs to see, so dropping them would defeat the report.
 */
fun summariseStudents(
    students: List<Student>,
    records: List<AttendanceRecord>,
    start: String,
    end: String,
): List<StudentSummary> {
    val inRange = records.filter { it.date in start..end }.groupBy { it.studentNumber }

    return students
        .filterNot { it.archived }
        .map { student -> StudentSummary(student, inRange[student.studentNumber].orEmpty().tally()) }
        .sortedBy { it.displayName }
}

data class DashboardEntry(
    val student: Student,
    val status: AttendanceStatus,
    val scheduleTitle: String,
    val recordedAt: String,
)

data class DashboardBreakdown(
    val entries: List<DashboardEntry>,
    val counts: StatusTally,
    /** Active students with no record at all yet for the chosen day. */
    val unaccountedFor: List<Student>,
)

/**
 * The day view, resolved back to real students.
 *
 * Records whose student has since been removed are dropped rather than rendered as an unknown
 * row: the record survives in storage for the audit trail, but a roster is a list of people.
 */
fun buildDashboard(
    students: List<Student>,
    records: List<AttendanceRecord>,
    date: String,
): DashboardBreakdown {
    val active = students.filterNot { it.archived }
    val byNumber = active.associateBy { it.studentNumber }

    val forDate = records.filter { it.date == date && it.studentNumber in byNumber }
    val entries = forDate
        .mapNotNull { record ->
            byNumber[record.studentNumber]?.let { student ->
                DashboardEntry(student, record.status, record.scheduleTitle, record.recordedAt)
            }
        }
        .sortedBy { it.student.displayName }

    val accountedFor = forDate.map { it.studentNumber }.toSet()

    return DashboardBreakdown(
        entries = entries,
        counts = forDate.tally(),
        unaccountedFor = active.filterNot { it.studentNumber in accountedFor },
    )
}

fun DashboardBreakdown.entriesWith(status: AttendanceStatus): List<DashboardEntry> =
    entries.filter { it.status == status }

fun List<DashboardEntry>.ofGender(gender: Gender): List<DashboardEntry> =
    filter { it.student.gender == gender }

File: web/src/domain/reporting.ts
import type { AttendanceRecord, AttendanceStatus, Gender, Student } from "./model";
import { fullName } from "./model";
import type { IsoDate } from "./primitives";
import { isWithinRange } from "./time";

export interface StatusTally {
  present: number;
  late: number;
  absent: number;
}

export const EMPTY_TALLY: Readonly<StatusTally> = Object.freeze({
  present: 0,
  late: 0,
  absent: 0,
});

export function tally(records: readonly AttendanceRecord[]): StatusTally {
  const result: StatusTally = { present: 0, late: 0, absent: 0 };
  for (const record of records) {
    result[record.status] += 1;
  }
  return result;
}

export function totalOf(counts: StatusTally): number {
  return counts.present + counts.late + counts.absent;
}

/**
 * Share of sessions the student actually turned up for. Late still counts as attending — it is
 * a punctuality problem, not an absence, and conflating the two was a real reporting bug in the
 * original app.
 */
export function attendanceRate(counts: StatusTally): number {
  const total = totalOf(counts);
  return total === 0 ? 0 : (counts.present + counts.late) / total;
}

export function formatRate(rate: number): string {
  return `${Math.round(rate * 100)}%`;
}

/* --------------------------------------------------------- Per-student report */

export interface StudentSummary {
  student: Student;
  displayName: string;
  counts: StatusTally;
  sessions: number;
  rate: number;
}

export interface SummaryRange {
  start: IsoDate;
  end: IsoDate;
}

/**
 * Attendance per student over an inclusive date range.
 *
 * Students with no records at all are still listed, with a zero tally — a student who never
 * appears in the data is exactly the one a coordinator needs to see, so dropping them would
 * defeat the report.
 */
export function summariseStudents(
  students: readonly Student[],
  records: readonly AttendanceRecord[],
  range: SummaryRange,
): StudentSummary[] {
  const byStudentNumber = new Map<string, AttendanceRecord[]>();
  for (const record of records) {
    if (!isWithinRange(record.date, range.start, range.end)) continue;
    const bucket = byStudentNumber.get(record.studentNumber);
    if (bucket) bucket.push(record);
    else byStudentNumber.set(record.studentNumber, [record]);
  }

  return students
    .filter((student) => !student.archived)
    .map((student) => {
      const counts = tally(byStudentNumber.get(student.studentNumber) ?? []);
      return {
        student,
        displayName: fullName(student),
        counts,
        sessions: totalOf(counts),
        rate: attendanceRate(counts),
      };
    })
    .sort((a, b) => a.displayName.localeCompare(b.displayName));
}

/* ------------------------------------------------------------ Daily dashboard */

export interface DashboardEntry {
  student: Student;
  displayName: string;
  status: AttendanceStatus;
  scheduleTitle: string;
  recordedAt: string;
}

export interface DashboardBreakdown {
  entries: DashboardEntry[];
  counts: StatusTally;
  /** Active students in the section with no record at all yet for the chosen day. */
  unaccountedFor: Student[];
}

/**
 * The day view: who was present, late, or absent, resolved back to real students.
 *
 * Records whose student has since been deleted are dropped rather than rendered as an unknown
 * row — the record survives in storage for the audit trail, but a roster is a list of people.
 */
export function buildDashboard(
  students: readonly Student[],
  records: readonly AttendanceRecord[],
  date: IsoDate,
): DashboardBreakdown {
  const activeStudents = students.filter((student) => !student.archived);
  const byNumber = new Map(activeStudents.map((student) => [student.studentNumber, student]));

  const forDate = records.filter((record) => record.date === date);
  const entries: DashboardEntry[] = [];
  const accountedFor = new Set<string>();

  for (const record of forDate) {
    const student = byNumber.get(record.studentNumber);
    if (!student) continue;
    accountedFor.add(record.studentNumber);
    entries.push({
      student,
      displayName: fullName(student),
      status: record.status,
      scheduleTitle: record.scheduleTitle,
      recordedAt: record.recordedAt,
    });
  }

  entries.sort((a, b) => a.displayName.localeCompare(b.displayName));

  return {
    entries,
    counts: tally(forDate.filter((record) => byNumber.has(record.studentNumber))),
    unaccountedFor: activeStudents.filter(
      (student) => !accountedFor.has(student.studentNumber),
    ),
  };
}

export function entriesByStatus(
  breakdown: DashboardBreakdown,
  status: AttendanceStatus,
): DashboardEntry[] {
  return breakdown.entries.filter((entry) => entry.status === status);
}

export function entriesByGender(
  entries: readonly DashboardEntry[],
  gender: Gender,
): DashboardEntry[] {
  return entries.filter((entry) => entry.student.gender === gender);
}
